### PR TITLE
Implemented NSIM similarity measurement.

### PR DESCRIPTION
**Pull request by zond**
- Implemented NSIM (DOI:10.1016/j.specom.2011.09.004).
- Replaced the geometric norm calculation across time and frequency axes with '1 - NSIM' when computing Zimtohrli distance.
- Removed the redundant cpp/zimt/README.md.
- Refactored the dataset management tools to enable optimization without saving the intermediary results.
- Implemented simple simulated annealing optimization for some relevant Zimtohrli parameters.
- Redid the MOS mapping so that Zimtohrli MOS scores are still on the same scale as before the NSIM change.

Before:

|Score type |MSE  |Min score |Max score |Mean score | 
|-----------|-----|----------|----------|-----------|
|ViSQOL     |0.09 |0.52      |0.80      |0.62       |
|PESQ       |0.10 |0.48      |0.84      |0.61       |
|Zimtohrli  |0.10 |0.56      |0.74      |0.59       |
|CDPAM      |0.12 |0.44      |0.73      |0.57       |
|PARLAQ     |0.15 |0.47      |0.78      |0.54       |
|PEAQB      |0.18 |0.28      |0.85      |0.51       |
|DPAM       |0.28 |0.19      |0.69      |0.42       |
|WARP-Q     |0.28 |0.07      |0.78      |0.45       |
|GVPMOS     |0.32 |0.01      |0.78      |0.40       |

After:

|Score type |MSE  |Min score |Max score |Mean score | 
|-----------|-----|----------|----------|-----------|
|ViSQOL     |0.09 |0.52      |0.80      |0.62       |
|Zimtohrli  |0.09 |0.59      |0.72      |0.60       |
|PESQ       |0.10 |0.48      |0.84      |0.61       |
|CDPAM      |0.12 |0.44      |0.73      |0.57       |
|PARLAQ     |0.15 |0.47      |0.78      |0.54       |
|PEAQB      |0.18 |0.28      |0.85      |0.51       |
|DPAM       |0.28 |0.19      |0.69      |0.42       |
|WARP-Q     |0.28 |0.07      |0.78      |0.45       |
|GVPMOS     |0.32 |0.01      |0.78      |0.40       |

See https://github.com/google/zimtohrli/pull/64/files?short_path=4640799#diff-46407991f9df1906835213b977238d969f72c49974a7dda6a9b1276daf199221 for the detailed performance changes.

### CODE REVIEW

File: cpp/zimt/nsim.cc [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+            channel_window_sum_data[channel_index];
+      }
+    }
+    // Populate tmp_a with windowed sums across steps-and-channels axes using
+    // the "prefix sums across the channel axis and windowed sums across the
+    // step axis" of tmp_b.
+    {
+      const float* step_prefix_sum_data = tmp_b[{step_index}].data();
+      float* step_window_sum_data = tmp_a[{step_index}].data();
+      for (size_t channel_index = 0; channel_index < channel_window;
```
Comment (szabadka) [inline cpp/zimt/nsim.cc:119]: this is a memcpy

File: cpp/zimt/nsim.cc [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+  // Populate tmp_b with prefix sums across the step axis.
+  {
+    float* channel_prefix_sum_data = tmp_b[{0}].data();
+    for (size_t channel_index = 0; channel_index < num_channels;
+         channel_index += Lanes(d)) {
+      Store(input_loader(0, channel_index), d,
+            channel_prefix_sum_data + channel_index);
+    }
+  }
+  for (size_t step_index = 1; step_index < num_steps; ++step_index) {
```
[the remark this answers is not in the document]
Comment (zond) [reply cpp/zimt/nsim.cc:64]: I did that until I realized that I had so many single variable loop variables that I started getting them mixed up - so I made the decision to keep them long and descriptive in this project.

File: cpp/zimt/nsim.h [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+#include <utility>
+#include <vector>
+
+#include "hwy/aligned_allocator.h"
+
+namespace zimtohrli {
+
+// Returns an array shaped exactly like source, where each element is the mean
+// of the step_window x channel_window rectangle of preceding elements.
```
[the remark this answers is not in the document]
Comment (zond) [reply cpp/zimt/nsim.h]: Done

File: cpp/zimt/nsim.h [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+hwy::AlignedNDArray<float, 2> WindowMean(
+    const hwy::AlignedNDArray<float, 2>& source, size_t step_window,
+    size_t channel_window);
+
+// Returns a slightly nonstandard version of the NSIM neural structural
+// similarity metric between arrays a and b.
```
[the remark this answers is not in the document]
Comment (zond) [reply cpp/zimt/nsim.h:33]: Done

File: cpp/zimt/nsim.cc [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+         channel_index += Lanes(d)) {
+      Store(Add(input_loader(step_index, channel_index),
+                Load(d, channel_prev_prefix_sum_data + channel_index)),
+            d, channel_prefix_sum_data + channel_index);
+    }
+  }
+
+  // Populate tmp_a with windowed sums across the step axis using the prefix
+  // sums in tmp_b.
+  for (size_t step_index = 0; step_index < step_window; ++step_index) {
```
[the remark this answers is not in the document]
Comment (zond) [reply cpp/zimt/nsim.cc]: Done

File: cpp/zimt/nsim.cc (hunk 1 above)
Comment (zond) [reply cpp/zimt/nsim.cc]: Done

**Pull request merged**